### PR TITLE
Allow NaN values to pass floating point bounds check.

### LINKED ISSUE #162: Deserialization: floating point bounds check does not permit messages with NaN
## Bug report

**Required Info:**

- Operating System: Linux 
- Installation type: Source, master
- Version or commit hash: 3dc587255effce330db05e045a5cead78df28339
- DDS implementation: Fast-RTPS
- Client library (if applicable): rclpy

#### Steps to reproduce issue
Running deserialize_message(serialized_message: bytes, message_type) on a message containing a field set to NaN triggers the bounds check added in #128.

assert value >= -3.402823e+38 and value <= 3.402823e+38,
AssertionError: The 'current' field must be a float in [-3.402823e+38, 3.402823e+38]

#### Expected behavior
The NaN value should be converted to a NaN in python.

The [BatteryState](https://github.com/ros2/common_interfaces/blob/b7e2ad503e1648468b1dc766c43aac12cc9aee5a/sensor_msgs/msg/BatteryState.msg) message, for example, uses NaNs to flag values that are not measured.

#### Actual behavior
The conversion triggers an assertion, and fails to deserialize.

### LINKED ISSUE #169: Serialization No Longer Allows float("inf") for float32 Type
In galactic release, we were able to use the numeric-limit value of float("inf") for a float32 type. When https://github.com/ros2/rosidl_python/pull/128 was introduced (and subsequently we began experimenting with humble release), we observed that the numeric limit checks prevented this use case. This is creating inconsistencies as our c++ code is still able to serialize using `std::numeric_limits<T>::infinity` but the python code is not able to do so. 

## Bug report

**Required Info:**

- Operating System:
  - ubuntu 22.04
- Installation type:
  - binaries
- Version or commit hash:
  - humble
- DDS implementation:
  - not relevant
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue
``` 
my_msg = Msg() # info field is float32
my_msg.info = float("inf")
```

#### Expected behavior
Set my_msg.info to INF (formerly would serialize to 3.402823e+40 in galactic)

#### Actual behavior
```
AssertionError: The 'info' field must be a float in [-3.402823e+38, 3.402823e+38]
```

### PR DESCRIPTION
**Pull request by oysstu**
This pull request resolves the issue #162 by explicitly adding NaN values to the floating point bounds check.

Signed-off-by: Oystein Sture <os@skarvtech.com>

Fixes #169
Fixes #162

### CODE REVIEW

**Comment by oysstu** [comment]
Math should only be imported for messages with floating point fields in order for it to pass the linter

**Comment by oysstu** [comment]
Updated the pull request. Inverted the bounds check instead of calling math.isnan. This works since NaN always evaluates to false in comparisons.

File: rosidl_generator_py/resource/_msg.py.em [hunk 1]
```diff
@@ -569,7 +569,7 @@ name = "double"
 bound = 1.7976931348623157e+308
 }@
 @[      end if]@
-            assert value >= -@(bound) and value <= @(bound), \
+            assert not (value < -@(bound) or value > @(bound)), \
```
Comment (sloretz) [inline rosidl_generator_py/resource/_msg.py.em:572]: Maybe this to allow inf too?

```
assert (not (value < -bound or value > bound)) or math.isinf(value)
```

Comment (sloretz) [inline rosidl_generator_py/resource/_msg.py.em:572]: Maybe this to allow inf too?

```python3
assert (not (value < -bound or value > bound)) or math.isinf(value)
```

File: rosidl_generator_py/resource/_msg.py.em [hunk 2]
```diff
@@ -506,14 +506,14 @@ bound = 2**nbits
 name = "float"
 bound = 3.402823e+38
 }@
-                 all(val >= -@(bound) and val <= @(bound) for val in value)), \
+                 not any(val < -@(bound) or val > @(bound) for val in value)), \
```
Comment (sloretz) [inline rosidl_generator_py/resource/_msg.py.em:509]: Maybe this to allow inf too?

```python3
all((not (value < -bound or value > bound)) or math.isinf(value) for value in array_value)
```

The typical case is all values are valid, so I like `all()` assuming it exits when it finds the first invalid value.

**Review by sloretz — DISMISSED** [review]
Thank you for the ticket and PR!

Would you be willing to update it to do two more things?
* Modify the comparison logic to allow positive and negative infinity (`-float('inf')` and `float('inf')`)
* Add tests for NaN, -inf, and inf for float32 and float64 types in [test_basic_types](https://github.com/ros2/rosidl_python/blob/6675823dd6123360912e233a830d1edab89c7651/rosidl_generator_py/test/test_interfaces.py#L40) and [test_arrays](https://github.com/ros2/rosidl_python/blob/6675823dd6123360912e233a830d1edab89c7651/rosidl_generator_py/test/test_interfaces.py#L292)

<details><summary>Playground script for myself when I re-review</summary>

```python3
import math


bound = 3.402823e+38


def old_float_check(value):
    return value >= -bound and value <= bound


def float_check(value):
    return not (value < -bound or value > bound)


def float_array_check(value):
    return not any(val < -bound or val > bound for val in value)


def alt_check(value):
    return (not (value < -bound or value > bound)) or math.isinf(value)


def alt_array_check(array_value):
    return all((not (value < -bound or value > bound)) or math.isinf(value) for value in array_value)


test_values = [
    float('nan'),
    float('inf'),
    -float('inf'),
    0,
    float(0),
    -float(0),
    3.14,
    -3.14,
]


print("New check")
for v in test_values:
    print(f'{float_check(v)} - {repr(v)}')
print(f'{repr(test_values)} - {float_array_check(test_values)}')


print ("Old check")
for v in test_values:
    print(f'{old_float_check(v)} - {repr(v)}')


print ("Alternative check")
for v in test_values:
    print(f'{alt_check(v)} - {repr(v)}')
print(f'{repr(test_values)} - {alt_array_check(test_values)}')
```
</details>

**Comment by oysstu** [comment]
In that case, the bounds check for float64 can be removed, as all numbers pass through it. Assigning a larger number than 1.7976931348623157e+308 to a regular python float or numpy.float64 automatically turns it into inf.

The bounds check may still make sense for float32, since there is no python builtin type, and may exceed the C/C++ float type. Question is if it should be converted to -inf/inf instead of throwing a bounds error, since this is the behavior for float64.

**Comment by connoranderson** [comment]
> Question is if it should be converted to -inf/inf instead of throwing a bounds error, since this is the behavior for float64.

Just my opinion here, I think silent overflow to inf makes sense for float32, if we do not have type-checking to validate that the user is actually providing a np.float32.

**Comment by oysstu** [comment]
> > Question is if it should be converted to -inf/inf instead of throwing a bounds error, since this is the behavior for float64.
> 
> Just my opinion here, I think silent overflow to inf makes sense for float32, if we do not have type-checking to validate that the user is actually providing a np.float32.

I agree. And that seems to be the behavior in IEEE 754 as far as I can tell. Honestly, I'm in favor of removing the entire bounds check for floating point numbers. I don't see what problem it solves.

**Comment by sloretz** [comment]
> The bounds check may still make sense for float32, since there is no python builtin type, and may exceed the C/C++ float type. Question is if it should be converted to -inf/inf instead of throwing a bounds error, since this is the behavior for float64.

> Honestly, I'm in favor of removing the entire bounds check for floating point numbers. I don't see what problem it solves.

The idea behind the checks using `@property` on the python classes was to make it easier to debug fields being set with invalid values. It raises an error at the point in code that a message was given an invalid value rather than getting an error when the message can't be serialized. The downside is these checks can be pretty expensive when setting a lot of fields. The checks can be skipped by [passing the -O option to cPython](https://docs.python.org/3/using/cmdline.html#cmdoption-O). Checking the bounds of `float32` here seems like it matches that intent.

Letting the conversion to `inf` happen means a message serialized and then deserialized might not be equal to itself, which feels non-intuitive to me.

```python3
from rclpy.serialization import serialize_message, deserialize_message

from test_msgs.msg import BasicTypes


def float32_value(self):
   return self._float32_value


def float32_value_setter(self, value):
   self._float32_value = value


BasicTypes.float32_value = property(float32_value, float32_value_setter)


if __name__ == '__main__':

    b = BasicTypes()
    b.float32_value = -3.402823e+40
    b_prime = deserialize_message(serialize_message(b), BasicTypes)

    print('b == b_prime?', b == b_prime)  # prints False because b_prime.float32_value is inf

```

As for `float64`, I agree the bounds check isn't needed as long as [sys.float_info.max and sys.float_info.min](https://docs.python.org/3/library/sys.html#sys.float_info) are the same as the bounds we'd be checking. If we get rid of the bounds check for float64 I'd still recommend a check that `sys.float_info` matches what we expect when the class is constructed.

**Comment by oysstu** [comment]
> Letting the conversion to `inf` happen means a message serialized and then deserialized might not be equal to itself, which feels non-intuitive to me.
I agree that this is undesirable. If the corresponding numpy type or python.ctypes type had been used in python, this would not be the case.

> As for `float64`, I agree the bounds check isn't needed as long as [sys.float_info.max and sys.float_info.min](https://docs.python.org/3/library/sys.html#sys.float_info) are the same as the bounds we'd be checking. If we get rid of the bounds check for float64 I'd still recommend a check that `sys.float_info` matches what we expect when the class is constructed.
That would mean that python is running on hardware that is not IEEE 754 compliant, or compiled with flags that disable conformance mode, but yes it can happen.

What if the floating point bounds checks are kept, but allows passing through variables explicitly set to Inf/NaN? How does that sound?

**Comment by sloretz** [comment]
> What if the floating point bounds checks are kept, but variables explicitly set to Inf/NaN are allowed through? How does that sound?

That sounds great to me

**Comment by oysstu** [comment]
I think the tests are failing because the assigned value gets rounded to inf and passes the bounds check, but I'll need to have a closer look at why it's happening. I won't have time to look at this again for a few days

File: rosidl_generator_py/test/test_interfaces.py [hunk 3]
```diff
@@ -460,9 +486,31 @@ def test_arrays():
     with pytest.raises(AssertionError):
         setattr(msg, 'uint64_values', [-1, 1, 2])
     with pytest.raises(AssertionError):
-        setattr(msg, 'float32_values', [-3.5e+38, 0.0, 3.5e+38])
-    with pytest.raises(AssertionError):
-        setattr(msg, 'float64_values', [-1.8e+308, 0.0, 1.8e+308])
+        float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
+        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0, float32_ieee_max_next])
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rosidl_generator_py/test/test_interfaces.py]: Addressed in 0431832

File: rosidl_generator_py/test/test_interfaces.py [hunk 4]
```diff
@@ -675,9 +723,15 @@ def test_bounded_sequences():
...
+
+    # If target system is IEEE 754 compliant, the next number is rounded to inf.
+    # Only perform this check on non-compliant systems.
+    if sys.float_info.max > 1.7976931348623157e+308:
+        with pytest.raises(AssertionError):
+            float64_ieee_max_next = numpy.nextafter(1.7976931348623157e+308, math.inf)
+            setattr(msg, 'float64_values', [-float64_ieee_max_next, 0.0, float64_ieee_max_next])
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rosidl_generator_py/test/test_interfaces.py:743]: Addressed in 0431832

File: rosidl_generator_py/test/test_interfaces.py [hunk 5]
```diff
@@ -818,9 +872,15 @@ def test_unbounded_sequences():
     with pytest.raises(AssertionError):
         setattr(msg, 'uint64_values', [-1, 1, 2])
     with pytest.raises(AssertionError):
-        setattr(msg, 'float32_values', [-3.5e+38, 0.0, 3.5e+38])
-    with pytest.raises(AssertionError):
-        setattr(msg, 'float64_values', [-1.8e+308, 0.0, 1.8e+308])
+        float32_ieee_max_next = numpy.nextafter(3.402823466e+38, math.inf)
+        setattr(msg, 'float32_values', [-float32_ieee_max_next, 0.0, float32_ieee_max_next])
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rosidl_generator_py/test/test_interfaces.py:885]: Addressed in 0431832

**Review by sloretz — APPROVED** [review]
LGTM with green CI!

**Comment by sloretz** [comment]
CI ([repos file](https://gist.githubusercontent.com/sloretz/6759feca2ad2365fca5742d2b098f9b6/raw/853bf6ff4e624c7975e67e6549d761013f8be5fe/ros2.repos) build: `--packages-above-and-dependencies rosidl_generator_py` test: `--packages-above rosidl_generator_py`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17144/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11691/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17568/)

**Comment by sloretz** [comment]
CI LGTM!

Thank you for the PR and for iterating!

**Comment by mmatthebi** [comment]
Hi, is there any plan when and if this PR will be merged into current Humble release?

**Comment by sloretz** [comment]
> Hi, is there any plan when and if this PR will be merged into current Humble release?

I'm cautious about backporting a behavior change to a low level package like this. If it is backported, then https://github.com/ros2/rosidl_python/pull/182 must be backported with it.

**Comment by mmatthebi** [comment]
> > Hi, is there any plan when and if this PR will be merged into current Humble release?
> 
> I'm cautious about backporting a behavior change to a low level package like this. If it is backported, then #182 must be backported with it.

thanks for the quick answer. In my eyes, this behaviour in Humble is a regression from Galactic, hence worth porting back into Humble. I actually wonder why this has not impeded more users. Our system relies on setting some message contents to `nan` to indicate that the value is empty and this is actually also commonly used in built-in messages (like [BatteryState](https://github.com/ros2/common_interfaces/blob/humble/sensor_msgs/msg/BatteryState.msg)). With the present behaviour we cannot move our setup to Humble LTS (which we would rather prefer).

**Comment by markcutler** [comment]
> > > Hi, is there any plan when and if this PR will be merged into current Humble release?
> > 
> > 
> > I'm cautious about backporting a behavior change to a low level package like this. If it is backported, then #182 must be backported with it.
> 
> thanks for the quick answer. In my eyes, this behaviour in Humble is a regression from Galactic, hence worth porting back into Humble. I actually wonder why this has not impeded more users. Our system relies on setting some message contents to `nan` to indicate that the value is empty and this is actually also commonly used in built-in messages (like [BatteryState](https://github.com/ros2/common_interfaces/blob/humble/sensor_msgs/msg/BatteryState.msg)). With the present behaviour we cannot move our setup to Humble LTS (which we would rather prefer).

+1 This is affecting our system as well. We would like to upgrade to Humble but need this PR backported first. Thanks.

**Comment by fujitatomoya** [comment]
@mergify backport humble

**Comment by fujitatomoya** [comment]
https://github.com/Mergifyio backport humble

**Comment by fujitatomoya** [comment]
@Mergifyio backport humble

**Review by fujitatomoya — COMMENTED** [review]
@clalancette @sloretz i think this can backport to humble, what do you think? (I do not have permission to manage backport though.)

**Pull request merged**